File: src/main/java/io/dev/mutex/thesisinfomgmt/service/AuthorService.java
package io.dev.mutex.thesisinfomgmt.service;

import io.dev.mutex.thesisinfomgmt.common.PaginatedData;
import io.dev.mutex.thesisinfomgmt.dto.AuthorDTO;

public interface AuthorService {

  /**
   * Retrieves authors
   *
   * @param page     the page of the data to be retrieved
   * @param pageSize the number of items to be retrieved
   * @param query    the query string
   * @return paginated authors data
   */
  PaginatedData<AuthorDTO> getAuthors(int page, int pageSize, String query);

  /**
   * Creates a new author
   *
   * @param author the details of the new author
   * @return created author
   */
  AuthorDTO createAuthor(AuthorDTO author);

  /**
   * Retrieves an author by its id
   *
   * @param id the id of the author to be retrieved
   * @return retrieved author
   */
  AuthorDTO getAuthor(long id);

  /**
   * Deletes an author
   *
   * @param id the id of the author to be deleted
   */
  void deleteAuthor(long id);

  /**
   * Updates an author
   *
   * @param id     the id of the author to be updated
   * @param author the updated author details
   */
  void updateAuthor(long id, AuthorDTO author);
}

File: app/src/services/AuthorService.ts
import { Author } from "../types/Authors";
import RequestHelper from "../helpers/RequestHelper";
import { PaginatedData } from "../types/Common";
import { RequestMethod } from "../common/constants";
import qs from "qs";

class AuthorService {
  resourcePath = "authors";

  async getAuthors(
    page: number,
    pageSize: number,
    query: string
  ): Promise<PaginatedData<Author>> {
    const queryParams = {
      page: page,
      pageSize: pageSize,
      query: query,
    };
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}?${qs.stringify(queryParams)}`,
      "get authors"
    );
  }

  async createAuthor(author: Author): Promise<Author> {
    return await RequestHelper.doRemoteCallAsync(
      this.resourcePath,
      "create author",
      RequestMethod.POST,
      author
    );
  }

  async updateAuthor(authorId: number, author: Author): Promise<void> {
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}/${authorId}`,
      "update author",
      RequestMethod.PUT,
      author
    );
  }

  async deleteAuthor(authorId: number): Promise<void> {
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}/${authorId}`,
      "delete author",
      RequestMethod.DELETE
    );
  }
}

export default new AuthorService();
